### PR TITLE
Added try-catch block while starting the watcherThread in ThreadDeath…

### PR DESCRIPTION
**Pull request by yaziciahmet**
ThreadDeathWatcher stops starting after some time, causing all the excess memory to be not cleaned, and causing a leak. It seems to be caused due to not handling the thread start exceptions, and started variable remains 1 even though the watcherThread is dead. I put it inside a try-catch block and converted it back to 0 if an exception occurs.

### CODE REVIEW

File: src/DotNetty.Common/ThreadDeathWatcher.cs [hunk 1]
```diff
@@ -59,10 +59,22 @@ static void Schedule(Thread thread, Action task, bool isWatch)
...
+                try
+                {
+                    var watcherThread = new Thread(s => ((IRunnable)s).Run());
+                    watcherThread.IsBackground = true;
+                    watcherThread.Start(watcher);
+                    ThreadDeathWatcher.watcherThread = watcherThread;
+                }
+                catch (Exception t)
+                {
+                    Logger.Warn("Thread death watcher raised an exception while trying to start the thread:", t);
```
Comment (LukaszRozmej) [inline src/DotNetty.Common/ThreadDeathWatcher.cs]: What is Logger here?

File: src/DotNetty.Common/ThreadDeathWatcher.cs [hunk 2]
```diff
@@ -59,10 +59,22 @@ static void Schedule(Thread thread, Action task, bool isWatch)
...
+                    if (!watcherThread.IsAlive)
+                    {
+                        bool stopped = Interlocked.CompareExchange(ref started, 0, 1) == 1;
+                        Contract.Assert(stopped);
```
Comment (LukaszRozmej) [inline src/DotNetty.Common/ThreadDeathWatcher.cs]: What is assert here?

File: src/DotNetty.Common/ThreadDeathWatcher.cs (hunk 1 above)
Comment (yaziciahmet) [reply src/DotNetty.Common/ThreadDeathWatcher.cs]: It is DotNetty's internal logger. They log all the errors / warnings using this logger.

File: src/DotNetty.Common/ThreadDeathWatcher.cs (hunk 2 above)
Comment (yaziciahmet) [reply src/DotNetty.Common/ThreadDeathWatcher.cs]: Assert to make sure that `started` was 1 to begin with. It is in my opinion impossible to get to this catch block and not having `started` as 1, but they also used this in another part of their code, so I wanted to use it as well.

Comment (yaziciahmet) [reply src/DotNetty.Common/ThreadDeathWatcher.cs]: Another part of their code where they convert `started` to 0.

Comment (LukaszRozmej) [reply src/DotNetty.Common/ThreadDeathWatcher.cs]: But will it log something? Throw exception?

File: src/DotNetty.Common/ThreadDeathWatcher.cs (hunk 1 above)
Comment (LukaszRozmej) [reply src/DotNetty.Common/ThreadDeathWatcher.cs]: Ok, do you know how to enable this dotnetty logs?

File: src/DotNetty.Common/ThreadDeathWatcher.cs (hunk 2 above)
Comment (yaziciahmet) [reply src/DotNetty.Common/ThreadDeathWatcher.cs]: It halts the program. "The default upgrade policy is to notify the attached debugger about the contract failure, or to display a Confirm dialog if the debugger is not attached"

**Comment by LukaszRozmej** [comment]
Cool is it used in https://github.com/NethermindEth/nethermind ?

**Pull request merged**